modules/javafx.base/src/main/java/javafx/collections/ModifiableObservableListBase.java

```diff
@@ -101,28 +101,46 @@ public boolean setAll(Collection<? extends E> col) {
 
     @Override
     public boolean addAll(Collection<? extends E> c) {
+        if (c.isEmpty()) {
+            return false;
+        }
+
         beginChange();
         try {
-            boolean res = super.addAll(c);
-            return res;
+            return super.addAll(c);
         } finally {
             endChange();
         }
     }
 
     @Override
     public boolean addAll(int index, Collection<? extends E> c) {
+        if (c.isEmpty()) {
+            if (index < 0 || index > size()) {
+                throw new IndexOutOfBoundsException("Index: " + index);
+            }
+
+            return false;
+        }
+
         beginChange();
         try {
-            boolean res = super.addAll(index, c);
-            return res;
+            return super.addAll(index, c);
         } finally {
             endChange();
         }
     }
 
     @Override
     protected void removeRange(int fromIndex, int toIndex) {
+        if (fromIndex == toIndex) {
+            if (fromIndex < 0 || fromIndex > size()) {
+                throw new IndexOutOfBoundsException("Index: " + fromIndex);
+            }
+
+            return;
+        }
```

Comment:
I'm really happy you also added the index range checks, not only does it conform to the contract better, it will help expose bugs in caller code (or even the JavaFX code as Observable* classes are used in many places, and who knows what bugs might lurk in some of the more complicated classes).

I'm not quite sure how this check is suppose to work though; does this need a comment ?
```suggestion
        if (fromIndex == toIndex) {  // distinguish between the clear and remove(int) call
            if (fromIndex < 0 || fromIndex > size()) {
                throw new IndexOutOfBoundsException("Index: " + fromIndex);
            }

            return;
        }
```

I think that would be a bit too cryptic for me...

Reply:
The index check is only really necessary for the optimized code path, since calling `listIterator` later in the method would catch an invalid index. However, I've opted to always check the index at the beginning of the method.